### PR TITLE
Release on jitpack.io

### PR DESCRIPTION
**Pull request by hannesa2**
I'm happy when someone tries it out 
https://jitpack.io/#ytai/ioio

Currently it builds only Android libraries, but maybe it's enough ?

image

### CODE REVIEW

**Comment by hannesa2** [comment]
Do we need versioninfo in `aar` as `BuildConfig.VERSIONNAME` ?

**Comment by topherbuckley** [comment]
@hannesa2 I'm not sure I understand the question. You want to remove the quoted code or replace the hard-coded definition of the version with a dynamic definition using the git commit count and tag? If the later, then that sounds like a good idea to me.  If so, wouldn't the versioninfo still be included in the aar?

I like the git integration you are going for, but how does the mlkit code you referenced handle the difference between releases and just snapshot or test builds? As it stands, with the`git describe --abbrev=0` it is my understanding that all subsequent commit info is dumped, and then you'd get identical versionName = TAG, for multiple builds on multiple commits. You'd see the difference in the commit count, but how would we identify which build was a release build and which was a WIP? In other words, I don't see how we'd differentiate between the release of v5.08 compared to some random WIP commit following the v5.08 release tag, but without any additional explicitly defined tag.  

Maybe my question stems from a misunderstanding of what the [getTag](https://github.com/hannesa2/mlkit/blob/6b20f89134d6ae82d9d895ea06ff3f46e6c24e07/android/mainSample/build.gradle#L102-L113) method is doing. What is the expected output of `$System.env.TAG_VERSION`, and where would this be defined or pulled from?

**Comment by topherbuckley** [comment]
>I'm happy when someone tries it out
https://jitpack.io/#ytai/ioio

I'm trying to integrate it into my project now. I replaced my local library

`api project(path: ':IOIOLibAndroid')`

with:

`api 'com.github.ytai:ioio:IOIOLibAndroid:6.0.0-beta3'`

and am getting the following error when trying to sync with gradle via Android Studio:

>Failed to resolve: com.github.ytai:ioio:IOIOLibAndroid

If I try running `./gradlew build` from the terminal, I get:

>FAILURE: Build failed with an exception.

>* What went wrong:
Could not determine the dependencies of task ':abcvlib:verifyReleaseResources'.
> Could not resolve all task dependencies for configuration ':abcvlib:releaseRuntimeClasspath'.
   > Could not find com.github.ytai:ioio:IOIOLibAndroid.
     Searched in the following locations:
       - file:/home/christopher/.m2/repository/com/github/ytai/ioio/IOIOLibAndroid/ioio-IOIOLibAndroid.pom
       - https://dl.google.com/dl/android/maven2/com/github/ytai/ioio/IOIOLibAndroid/ioio-IOIOLibAndroid.pom
       - https://jcenter.bintray.com/com/github/ytai/ioio/IOIOLibAndroid/ioio-IOIOLibAndroid.pom
       - https://jitpack.io/com/github/ytai/ioio/IOIOLibAndroid/ioio-IOIOLibAndroid.pom

I'll work on this a bit today on my fork to see if I can get it to resolve somehow. 

------------------------------

>Currently it builds only Android libraries, but maybe it's enough ?

I also implement IOIOLibCore locally in my project at the moment, so I'd like some way to package this as well. I'll work on it today a bit.

**Comment by hannesa2** [comment]
> Failed to resolve: com.github.ytai:ioio:IOIOLibAndroid

It sounds like you forgot 

	allprojects {
		repositories {
			...
			maven { url 'https://jitpack.io' }
		}
	}

as I mentioned in this pull request

> I also implement IOIOLibCore locally in my project at the moment, so I'd like some way to package this as well. I'll work on it today a bit.

You don't need to include `IOIOLibCore` in Android projects, because it should be [included](https://github.com/ytai/ioio/blob/master/software/IOIOLibAndroid/build.gradle#L19-L22) already

**Comment by topherbuckley** [comment]
>It sounds like you forgot
```
allprojects {
	repositories {
		...
		maven { url 'https://jitpack.io' }
	}
}
```
Nope, thats in my application (root) project folder and I also tried with it in the module level gradle file as well. 

>You don't need to include IOIOLibCore in Android projects, because it should be included already

Good point. I think I originally had to include IOIOLIbCore in my project since I believe that the `api` call used to be `implementation`. And had overlooked the doubled dependency.

**Comment by topherbuckley** [comment]
I may have had the maven url repo in the wrong "repositories" location. I have multiple. I added it to the buildscript->repositories section as well. 

Here is the root project build script:

```groovy
buildscript {
    repositories {
        google()
        jcenter()
        maven { url 'https://jitpack.io' }
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:4.0.1'
        classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.72'
    }
}

allprojects {
    repositories {
        mavenLocal() //Todo not sure if I need this
        google()
        jcenter()
        maven { url 'https://jitpack.io' }
    }
    configurations {
        all {
            exclude module: 'commons-logging'
        }
    }
}

subprojects {
    if (project.name != 'IOIOLibCore'){
        if (gradle.ext.androidLibs.any{name -> name == project.name}){
            apply plugin: 'com.android.library'
        } else if (gradle.ext.apps.any{name -> name == project.name}){
            apply plugin: 'com.android.application'
            dependencies {
                implementation(project(":abcvlib"))

                implementation 'com.google.code.gson:gson:2.8.5'

                implementation 'androidx.annotation:annotation:1.1.0'
                implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
                implementation "androidx.appcompat:appcompat:1.2.0"
                implementation "androidx.lifecycle:lifecycle-livedata:2.2.0"
                implementation "androidx.lifecycle:lifecycle-viewmodel:2.2.0"
            }
        }
        android{
            compileOptions {
                sourceCompatibility(JavaVersion.VERSION_1_8)
                targetCompatibility(JavaVersion.VERSION_1_8)
            }
            defaultConfig {
                compileSdkVersion = 28
                buildToolsVersion = '30.0.2'
                minSdkVersion 21
                targetSdkVersion 28
                versionCode = 1
                versionName = "1.0"
            }
        }
    }
}

task clean(type: Delete) {
    delete rootProject.buildDir
}
```

and the module that is failing:

```groovy
dependencies {
//    api project(path: ':IOIOLibAndroid')
//    api project(path: ':IOIOLibCore')
    api 'com.github.ytai:ioio:IOIOLibAndroid:6.0.0-beta3'
    implementation 'com.opencsv:opencsv:4.6'
    api project(path: ':openCV411')
}
```

and my settings.gradle

```groovy
def androidLibs = [
        'abcvlib',
        'openCV411',
        'IOIOLibAndroid',
        'IOIOLibAndroidAccessory',
        'IOIOLibAndroidBluetooth',
        'IOIOLibAndroidDevice'
]

gradle.ext.androidLibs = androidLibs

def apps = [
        'basic',
        'camera',
        'backAndForth',
        'pidBalancer',
        'colorBlobFollower',
        'setPath',
        'clapLearn',
        'compoundController',
        'allAccess',
        'physicstest',
]

gradle.ext.apps = apps

def IOIOlibs = [
        'IOIOLibCore'
]

gradle.ext.IOIOlibs = IOIOlibs

androidLibs.each {
    lib ->
        include lib
        project(":${lib}").projectDir = file("libs/${lib}")
}

apps.each {
    app ->
        include app
        project(":${app}").projectDir = file("apps/${app}")
}

IOIOlibs.each {
    lib ->
        include lib
        project(":${lib}").projectDir = file("libs/${lib}")
}

```

Its a bit of a mess still, but maybe this points to something I'm not seeing. Also you might see the logic I'm using to to package up all my modules in settings.gradle. We might find this useful for eliminating a few more boilerplate lines form the module-level build scripts in this library as well.

**Comment by topherbuckley** [comment]
Spotted it! Just a typo!

`    api 'com.github.ytai:ioio:IOIOLibAndroid:6.0.0-beta3'`

should have been

`    api 'com.github.ytai.ioio:IOIOLibAndroid:6.0.0-beta3'`

**Comment by topherbuckley** [comment]
It looks like this should work, but the build was not based on the appCompatActivity branch, so it will fail due to the MLKit sample code dependency on this. Can you rebase the Jitpack branch on master and rebuild?

**Comment by hannesa2** [comment]
> so it will fail due to the MLKit sample code dependency on this.

Sorry, I missed what the _MLKit sample code dependency_ is

**Comment by topherbuckley** [comment]
No no, I was not clear, sorry. 

## TLDR:
I use [CameraXLivePreviewActivity](https://github.com/googlesamples/mlkit/blob/e949654bef09e97b237c7692349c52ee06e9fd9e/android/vision-quickstart/app/src/main/java/com/google/mlkit/vision/demo/java/CameraXLivePreviewActivity.java#L83) as the base class for a module in my project. I need IOIOActivity to extend AppCompatActivity rather than Activity for my project to work. 

## Details:

So the project that I use this library in has some modules that fork from the MLKit vision-quickstart sample code. In particular, I use the [CameraXLivePreviewActivity](https://github.com/googlesamples/mlkit/blob/e949654bef09e97b237c7692349c52ee06e9fd9e/android/vision-quickstart/app/src/main/java/com/google/mlkit/vision/demo/java/CameraXLivePreviewActivity.java#L83) as the base class for my module. I added some of my own code to the sample (e.g. what to do when an object is detected, etc.), and changed the inheritance from AppCompatActivity to the wrapper (AbcvlibActivty), which extends IOIOActivity so that I can have both the MLKit vision functionality as well as the IOIO functionality (both required for my use-case). 

Initially I was just implementing various interfaces into my wrapper to make up for the differences between Activity and AppCompatActivity (things like LifeCycle, among others I don't recall off hand), but this was getting bulky and was an uphill battle. 

All of my issues sum up to trying to work around Java's single inheritance limitation. I could either run the IOIOLib stuff as a service in the background and use the example MLKit code as is, or the opposite. Either way would cost a LOT of time to rewrite my existing code base.  

The easiest way to get around this was to just have a local version of IOIOLibCore in my project and change out the inheritance from Activity to AppCompatActivity. This worked well enough, but it required I do all the initial refactoring I did to bring this library up to speed with Android Studio and the SDK 29. Thus what brought me here as a contributor.

**Comment by hannesa2** [comment]
now you can try `6.0.0-beta4`

**Comment by hannesa2** [comment]
I made a test pull request for jitpack.io usage #174

**Comment by topherbuckley** [comment]
>now you can try 6.0.0-beta4

Tested and built successfully. I'll test it out in my hardware in a bit and let you know if anything seems off.

**Comment by topherbuckley** [comment]
Following up just to say that I tested it on my hardware, and my robot is showing no signs of disliking the new build. :smile:

File: IOIOLibAndroid/build.gradle [hunk 1]
```diff
@@ -1,16 +1,5 @@
...
-        mavenDeployer {
-            pom.project {
-                name = 'IOIOLibAndroid'
-                description = 'IOIO library for Android - common code with ADB support'
-                packaging = 'aar'
-            }
-        }
-    }
-}
+apply plugin: 'com.github.dcendents.android-maven'
```
Comment (topherbuckley) [inline IOIOLibAndroid/build.gradle:2]: I'm confused about this. It looks like this project [was abandoned](https://github.com/dcendents/android-maven-gradle-plugin). What is this for?

File: software/IOIOLibAndroid/build.gradle [hunk 2]
```diff
@@ -1,16 +1,7 @@
 apply plugin: 'com.android.library'
+apply plugin: 'com.github.dcendents.android-maven'
 
-uploadArchives {
-    repositories {
-        mavenDeployer {
-            pom.project {
```
[the remark this answers is not in the document]
Comment (hannesa2) [reply software/IOIOLibAndroid/build.gradle]: I simplified it. 
If you see a benefit somewhere (and at least working and used) go for it !

File: IOIOLibAndroid/build.gradle (hunk 1 above)
Comment (hannesa2) [reply IOIOLibAndroid/build.gradle:2]: To make jitpack work

File: software/IOIOLibAndroid/build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (topherbuckley) [reply software/IOIOLibAndroid/build.gradle]: The stuff I see being lost is the description and details given in the past [maven repo](https://mvnrepository.com/artifact/com.github.ytai.ioio). If you look into a single lib, single version, you see the pom info including [ytai's email](https://mvnrepository.com/artifact/com.github.ytai.ioio/IOIOLibAndroidBluetooth/5.07). This can be an initial point of contact, but I can't claim to ever have used such a path. 

I don't see any similar POM info available on jitpack, but maybe the dev process of using jitpack is just so different that its not necessary? If someone is using the jitpack builds, then they are likely getting there from here (github), and if that is the case, then all that info is already available here. 

On the other hand, if someone finds the jitpack builds somewhere else (e.g. some sample code that imports the ioio jitpack lib), then the first thing they'd see is a description-less version tag and nothing about what the library represents. I guess I see a possibility for confusion, but the likelihood of it actually being a problem VERY small, so I agree, this can be neglected unless someone says otherwise.

File: IOIOLibAndroid/build.gradle (hunk 1 above)
Comment (topherbuckley) [reply IOIOLibAndroid/build.gradle:2]: Sorry, yes. I had forgotten [this](https://jitpack.io/docs/ANDROID/). Thought this was something new for some reason.
